rust/builder/mod.rs

```diff
@@ -33,83 +33,81 @@ use crate::{
 
 #[macro_export]
 macro_rules! typeql_match {
-    ($($pattern:expr),* $(,)?) => {{
-        $crate::query::MatchClause::from_patterns(vec![$($pattern.into()),*])
-    }}
+    ($($pattern:expr),* $(,)?) => {
+        $crate::query::MatchClause::new($crate::pattern::Conjunction::new(vec![$($pattern.into()),*]))
+    }
 }
 
 #[macro_export]
 macro_rules! typeql_insert {
-    ($($thing_statement:expr),* $(,)?) => {{
+    ($($thing_statement:expr),* $(,)?) => {
         $crate::query::TypeQLInsert::new(vec![$($thing_statement),*])
-    }}
+    }
 }
 
 #[macro_export]
 macro_rules! typeql_define {
-    ($($pattern:expr),* $(,)?) => {{
+    ($($pattern:expr),* $(,)?) => {
         $crate::query::TypeQLDefine::new(vec![$($pattern.into()),*])
-    }}
+    }
 }
 
 #[macro_export]
 macro_rules! typeql_undefine {
-    ($($pattern:expr),* $(,)?) => {{
+    ($($pattern:expr),* $(,)?) => {
         $crate::query::TypeQLUndefine::new(vec![$($pattern.into()),*])
-    }}
+    }
 }
 
 #[macro_export]
 macro_rules! and {
-    ($($pattern:expr),* $(,)?) => {{
+    ($($pattern:expr),* $(,)?) => {
         $crate::pattern::Conjunction::new(vec![$($pattern.into()),*])
-    }}
+    }
 }
 
 #[macro_export]
 macro_rules! or {
-    ($($pattern:expr),* $(,)?) => {{
-        let mut patterns = vec![$($pattern.into()),*];
-        match patterns.len() {
-            1 => patterns.pop().unwrap(),
-            _ => $crate::pattern::Disjunction::new(patterns).into(),
-        }
-    }}
+    ($pattern:expr $(,)?) => { 
+        compile_error!("Useless disjunction of one pattern");
+    };
+
+    ($($pattern:expr),+ $(,)?) => {
+        $crate::pattern::Disjunction::new(vec![$($pattern.into()),*])
+    };
 }
```

Comment:
This change allows us to actually return a `Disjunction` rather than a generic `Pattern`, and avoids confusion where `or!(x)` would simply return `x` rather than a disjunction containing `x`.